fish-mode.el

```diff
@@ -335,18 +338,18 @@ For example, (fold F X '(1 2 3)) computes (F (F (F X 1) 2) 3)."
       (setq x2 (funcall f x2 (pop li))))
     x2))
 
-(defun fish/count-of-tokens-in-string (token token-to-ignore string)
-  (let ((count 0)
-        (pos 0))
-    (while pos
-      (if (and token-to-ignore
-               (string-match token-to-ignore string pos))
-          (setq pos (match-end 0)))
-      (if (string-match token string pos)
-          (setq pos (match-end 0)
-                count (+ count 1))
-        (setq pos nil)))
-    count))
+(defun fish/count-tokens-on-current-line (positive-re negative-re)
+  "Return count of matches for POSITIVE-RE that do not also match NEGATIVE-RE on current line.
+POSITIVE-RE and NEGATIVE-RE are regular expressions."
+  (save-excursion
+    (cl-loop initially (goto-char (line-beginning-position))
+             while (re-search-forward positive-re (line-end-position) t)
+             count (not (or (when negative-re
+                              (looking-at-p negative-re))
+                            ;; String
+                            (nth 3 (syntax-ppss))
+                            ;; Comment
+                            (nth 4 (syntax-ppss)))))))
```

Comment:
I wondered about that too.  I haven't noticed any problems, but I looked at `syntax-ppss`, and it is a rather large function, so I'll capture it in a variable like you suggested.